src/updates.rs

```diff
@@ -0,0 +1,128 @@
+/// ignored in release builds.
+pub(crate) fn begin_checking_for_updates() {
+    let result = try_begin_checking_for_updates();
+    if cfg!(debug_assertions) {
+        result.unwrap();
+    }
+}
+
+fn try_begin_checking_for_updates() -> io::Result<()> {
+    let path = path_to_latest_version_file()?;
+    let time_since_modified = time_since_last_update(&path)?;
+    if time_since_modified >= ONE_DAY {
+        // Eagerly update the timestamp. If something goes wrong, we don't
+        // want to be spamming GitHub's API.
+        filetime::set_file_mtime(&path, filetime::FileTime::now())?;
+        let mut command = Command::new(env::current_exe()?);
+        command.arg("--internal-fetch-latest-version");
+        #[cfg(not(debug_assertions))]
+        {
+            command.stdout(Stdio::null());
+            command.stderr(Stdio::null());
+        }
+        command.spawn()?;
+    }
+    Ok(())
+}
+
+/// Reports if updates are available
+///
+/// This will check a file which contains the latest available version of rubyfmt. If it is greater
+/// than the current version, we will print to stderr that an update is available. Since this
+/// process is strictly optional, any errors will be silently ignored.
+///
+/// For most code bases, rubyfmt will finish running well before the update check is completed.
+/// This means that this function will most likely only print a warning on the second run after an
+/// update is available.
+pub(crate) fn report_if_update_available() {
+    let result = try_report_if_update_available();
+    if cfg!(debug_assertions) {
+        result.unwrap();
+    }
+}
+
+fn try_report_if_update_available() -> io::Result<()> {
+    let latest_version = latest_available_rubyfmt_version()?;
+    if latest_version > installed_rubyfmt_version() {
+        eprintln!("A new version of rubyfmt is available at https://github.com/penelopezone/rubyfmt/releases/tag/v{}", latest_version);
+    }
+    Ok(())
+}
+
+fn installed_rubyfmt_version() -> semver::Version {
+    semver::Version::parse(env!("CARGO_PKG_VERSION"))
+        .expect("$CARGO_PKG_VERSION should always be a valid semver version")
+}
+
+fn latest_available_rubyfmt_version() -> io::Result<semver::Version> {
+    let latest_version_str = fs::read_to_string(path_to_latest_version_file()?)?;
+    semver::Version::parse(&latest_version_str)
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+}
+
+/// Determine the most recent version of Rubyfmt, and write it to a file
+///
+/// Because GitHub paginates its responses to this endpoint, this assumes that the latest version
+/// is in the first 30 tags given to us (which are sorted in lexographically descending order).
+/// Tags which are a version release are assumed to always be `v` followed by a semver version (for
+/// example, v1.0.0)
+pub(crate) fn fetch_latest_version() -> Result<(), Box<dyn Error>> {
+    let max_version = ureq::get("https://api.github.com/repos/penelopezone/rubyfmt/tags")
```

Comment:
can this be releases instead of tags?

Reply:
I'll check. If we can do you want it to always be the latest release or should I continue parsing the name and grab the highest semver version as insurance against releasing a security patch for an old version

Reply:
So it looks like I *can* hit the releases endpoint instead of tags, but I'm just going to have to end up using the tag name anyway since the release name hasn't been consistent. I don't think there's any benefit to doing this since the releases endpoint is effectively just the tags endpoint with a much larger payload size in this case, but I'm happy to write the code if that's what you want

Reply:
cool, I think this is fine as is 🚀